core/src/main/java/org/apache/iceberg/encryption/KeyMetadata.java

```diff
@@ -0,0 +1,148 @@
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.apache.iceberg.encryption;
+
+import static org.apache.iceberg.types.Types.NestedField.optional;
+import static org.apache.iceberg.types.Types.NestedField.required;
+
+import java.io.IOException;
+import java.nio.ByteBuffer;
+import org.apache.avro.generic.IndexedRecord;
+import org.apache.iceberg.Schema;
+import org.apache.iceberg.avro.AvroSchemaUtil;
+import org.apache.iceberg.data.GenericRecord;
+import org.apache.iceberg.data.Record;
+import org.apache.iceberg.types.Types;
+
+public class KeyMetadata implements EncryptionKeyMetadata, IndexedRecord {
+  private static final ThreadLocal<KeyMetadataEncoder<Record>> ENCODER = new ThreadLocal<>();
+  private static final ThreadLocal<KeyMetadataDecoder<Record>> DECODER = new ThreadLocal<>();
+
+  private static final String encryptionKeyField = "encryption_key";
+  private static final String wrappingKeyIdField = "wrapping_key_id";
+  private static final String aadPrefixField = "aad_prefix";
+
+  static final byte V1 = 1;
+  private static final Schema SCHEMA_V1 =
+      new Schema(
+          required(0, encryptionKeyField, Types.BinaryType.get()),
+          optional(1, wrappingKeyIdField, Types.StringType.get()),
+          optional(2, aadPrefixField, Types.BinaryType.get()));
+
+  private String wrappingKeyId;
+  private ByteBuffer encryptionKey;
+  private ByteBuffer aadPrefix;
+
+  KeyMetadata(ByteBuffer encryptionKey, String wrappingKeyId, ByteBuffer aadPrefix) {
+    this.wrappingKeyId = wrappingKeyId;
+    this.encryptionKey = encryptionKey;
+    this.aadPrefix = aadPrefix;
+  }
+
+  public String wrappingKeyId() {
+    return wrappingKeyId;
+  }
+
+  public ByteBuffer encryptionKey() {
+    return encryptionKey;
+  }
+
+  public ByteBuffer aadPrefix() {
+    return aadPrefix;
+  }
+
+  public static KeyMetadata parse(ByteBuffer buffer) {
+    KeyMetadataDecoder<Record> decoder = DECODER.get();
+    if (decoder == null) {
+      decoder = new KeyMetadataDecoder<>(V1, SCHEMA_V1);
+      DECODER.set(decoder);
+    }
+    try {
+      Record rec = decoder.decode(buffer);
+      return new KeyMetadata(
+          (ByteBuffer) rec.getField(encryptionKeyField),
+          (String) rec.getField(wrappingKeyIdField),
+          (ByteBuffer) rec.getField(aadPrefixField));
+    } catch (IOException e) {
+      throw new RuntimeException("Failed to parse envelope encryption metadata", e);
+    }
+  }
+
+  @Override
+  public ByteBuffer buffer() {
+    Record rec = GenericRecord.create(SCHEMA_V1.asStruct());
+    rec.setField(encryptionKeyField, encryptionKey);
+    rec.setField(wrappingKeyIdField, wrappingKeyId);
+    rec.setField(aadPrefixField, aadPrefix);
```

Comment:
This should not copy the record. It should directly encode or decode because this implements `IndexedRecord`